File: app/src/main/java/core/Password.java
package core;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Password {
    private int id;
    private int position;
    private String username;
    private String program;
    private List<PasswordHistory> passwordHistory;


    public Password(int id, int position, String username, String program) {
        this.id = id;
        this.position = position;
        this.username = username;
        this.program = program;
        this.passwordHistory = new ArrayList<>();
    }

    public void logout() {
        id = Integer.MIN_VALUE;
        position = Integer.MIN_VALUE;
        username = null;
        program = null;
        passwordHistory.clear();
        passwordHistory = null;
    }

    void addHistoryItem(int id, String value, String date) {
        DateFormat dateFormat = new SimpleDateFormat();
        Date convertedDate = new Date();
        try {
            convertedDate = dateFormat.parse(date);
        } catch (ParseException e) {
            e.printStackTrace();
        }

        PasswordHistory history = new PasswordHistory(id, value, convertedDate);
        passwordHistory.add(history);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Password)) return false;

        Password password = (Password) o;

        if (id != password.id) return false;
        if (position != password.position) return false;
        if (!username.equals(password.username)) return false;
        return program.equals(password.program);

    }

    @Override
    public int hashCode() {
        int result = id;
        result = 31 * result + position;
        result = 31 * result + username.hashCode();
        result = 31 * result + program.hashCode();
        return result;
    }

    public int getId() {
        return id;
    }

    public int getPosition() {
        return position;
    }

    public String getUsername() {
        return username;
    }

    public String getProgram() {
        return program;
    }

    public List<PasswordHistory> getPasswordHistory() {
        return passwordHistory;
    }

    public PasswordHistory getFirstItem() {
        return passwordHistory.get(0);
    }

    @Override
    public String toString() {
        return "Password{" +
                "username='" + username + '\'' +
                "password='" + getFirstItem().getValue() + '\'' +
                ", program='" + program + '\'' +
                '}';
    }
}

File: app/src/main/java/com/typingsolutions/kore/common/data/Password.kt
package com.typingsolutions.kore.common.data

import net.sqlcipher.database.SQLiteDatabase
import java.text.SimpleDateFormat
import java.util.*

data class Password(val id: Int, val username: String, val program: String, val position: Int, val passwords: ArrayList<History>) {
    override fun equals(other: Any?): Boolean {
        if(other == null) return false
        if(other.javaClass != javaClass) return false

        return (other as Password).id == id
    }

    override fun hashCode(): Int = id

    companion object {
        val CreateString: String = "CREATE TABLE IF NOT EXISTS passwords (id INTEGER PRIMARY KEY, mUsernameTextView TEXT, mProgramTextView TEXT, position INT)"

        fun loadPasswords(database: SQLiteDatabase): Array<Password> {
            val result = ArrayList<Password>()
            val query = "SELECT p.id, " +
                               "p.mUsernameTextView, " +
                               "p.mProgramTextView, " +
                               "p.position, " +
                               "h.id, " +
                               "h.mPasswordTextView, " +
                               "h.created " +
                        "FROM passwords p " +
                        "JOIN history h ON p.id = h.passwordId " +
                        "ORDER BY p.position, p.id, h.created"
            val cursor = database.rawQuery(query, arrayOf())

            var last: Password? = null

            while(cursor.moveToNext())
            {
                val id = cursor.getInt(0)
                val username = cursor.getString(1)
                val program = cursor.getString(2)
                val position = cursor.getInt(3)
                val historyId = cursor.getInt(4)
                val password = cursor.getString(5)
                val createString = cursor.getString(6)

                val df = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

                val item = History(historyId, password, df.parse(createString))
                val list = ArrayList<History>()
                list.add(item)
                val current = Password(id, username, program, position, list)

                if (current == last) {
                    last.passwords.add(item)
                } else {
                    if(last != null) {
                        result.add(last)
                    }
                    last = current
                }

            }

            cursor.close()

            return result.toTypedArray()
        }
    }
}